controller.go

```diff
@@ -170,60 +172,80 @@ func (n *NodeController) nodeReady(node *v1.Node) (bool, error) {
 // setNodeReady sets node taint macthing ready value. E.g. sets NotReady taint
 // if ready is false, and removes the taint (if exists) when ready is true.
 func (n *NodeController) setNodeReady(node *v1.Node, ready bool) error {
-	// if ready, remove notReady taint if exists on the node
-	if ready {
-		var newTaints []v1.Taint
-		for _, taint := range node.Spec.Taints {
-			if taint.Key != n.taintNodeNotReadyName {
-				newTaints = append(newTaints, taint)
-			}
+	setNodeReadiness := func() error {
+		updatedNode, err := n.CoreV1().Nodes().Get(node.Name, metav1.GetOptions{})
+		if err != nil {
+			return backoff.Permanent(err)
 		}
 
-		if len(newTaints) != len(node.Spec.Taints) {
-			node.Spec.Taints = newTaints
-			_, err := n.CoreV1().Nodes().Update(node)
-			if err != nil {
-				return err
-			}
-			log.WithFields(log.Fields{
-				"action": "removed",
-				"taint":  n.taintNodeNotReadyName,
-				"node":   node.ObjectMeta.Name,
-			}).Info("")
-
-			if n.nodeStartUpObeserver != nil {
-				// observe node startup duration
-				n.nodeStartUpObeserver.ObeserveNode(*node)
+		// if ready, remove notReady taint if exists on the node
+		if ready {
+			var newTaints []v1.Taint
+			for _, taint := range updatedNode.Spec.Taints {
+				if taint.Key != n.taintNodeNotReadyName {
+					newTaints = append(newTaints, taint)
+				}
 			}
 
-			// trigger hooks on node ready.
-			for _, hook := range n.nodeReadyHooks {
-				err := hook.Trigger(node.Spec.ProviderID)
+			if len(newTaints) != len(updatedNode.Spec.Taints) {
+				updatedNode.Spec.Taints = newTaints
+				_, err := n.CoreV1().Nodes().Update(updatedNode)
 				if err != nil {
-					log.Errorf("Failed to trigger hook '%s': %v", hook.Name(), err)
+					// automatically retry if there was a conflicting update.
+					serr, ok := err.(*apiErrors.StatusError)
```

Comment:
I think you can do `errors.IsConflict` here (and I should fix it in CLM as well!).